Frontend/save schedules

ryan-conn opened the pull request:
## Description

Adds functionality to save/unsave and remove schedules.

## Rationale

A couple of potentially controversial/questionable changes I made:
- Colors for save/delete buttons are the same because I wasn't sure what color if any to change them to, if anyone has suggestions then go ahead
- A couple tests are causing a minor error I'm not sure how to fix, I'll post a comment describing this in more detail

## How to test

Just generate schedules and save them, you can see that when you click generate schedules again it keeps your saved schedules.

## Screenshots

|Before|After|
|--|--|
|![image](https://user-images.githubusercontent.com/28655462/84580219-94b9e100-ad9a-11ea-976d-646c2298d5e5.png)|![image](https://user-images.githubusercontent.com/28655462/86314811-38eeb500-bbee-11ea-8b0b-fdb32b0581a7.png)|

## Related tasks

Closes #228.

ryan-conn commented on autoscheduler/frontend/src/tests/ui/SchedulePreview.test.tsx:
```diff
@@ -145,4 +147,94 @@ describe('SchedulePreview component', () => {
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      // save, wait for save to finish, then unsave
+      const saveScheduleButton = (await findAllByText('Save'))[0];
+      fireEvent.click(saveScheduleButton);
+      await waitFor(() => getByText('Unsave'));
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toHaveLength(0)
+      ));
+    });
+
+    test('when a schedule with index greater than 0 is unsaved', async () => {
+      // arrange
+      const store = createStore(autoSchedulerReducer);
+      const { findAllByText, getByText } = render(
+        <Provider store={store}>
+          <SchedulePreview />
+        </Provider>,
+      );
+      store.dispatch(replaceSchedules([testSchedule1, testSchedule2]));
+
+      // act
+      // save, wait for save to finish, then unsave
+      const saveScheduleButton = (await findAllByText('Save'))[1];
+      fireEvent.click(saveScheduleButton);
+      await waitFor(() => getByText('Unsave'));
+      fireEvent.click(saveScheduleButton);
+
+      // assert
+      await waitFor(() => (
+        expect(store.getState().schedules.savedSchedules).toHaveLength(0)
+      ));
+    });
```
These tests cause the following error:
![image](https://user-images.githubusercontent.com/28655462/84580399-3f7ecf00-ad9c-11ea-8b94-a6b1b83259ec.png)
This is because of the line `await waitFor(() => getByText('Unsave'));` (which I realize could just be `await findByText('Unsave')`, not sure why I didn't do that in hindsight but it still results in the same error). This error doesn't occur when actually using the UI so it's pretty minor and the tests still work correctly despite it but still, it would be nice to fix.

firejake308 replied:
I would move the `sleep` function to `testUtil.ts` and mock it to return immediately during the test

ryan-conn replied:
I just removed the sleep function since everyone's seen how it looks and we won't want it in master anyways

firejake308 commented on autoscheduler/frontend/src/redux/reducers/schedules.ts:
```diff
@@ -22,18 +24,77 @@ export interface ReplaceSchedulesAction {
+
+interface Schedules {
+  schedules: Meeting[][];
+  savedSchedules: Meeting[][];
```
Doesn't this result in duplication of data? I'm thinking `savedSchedules` should be like an array of indices

ryan-conn replied:
No, objects in arrays are references to the objects, not the actual data so storing it as a number or a reference take up the same memory, as both are 8 bytes

firejake308 commented on autoscheduler/frontend/src/redux/reducers/schedules.ts:
```diff
@@ -22,18 +24,77 @@ export interface ReplaceSchedulesAction {
+}
+
+const initialSchedules: Schedules = {
+  schedules: [],
```
Maybe rename this to `allSchedules` or something to be more specific about what it's doing

ryan-conn replied:
Good idea, referencing `schedules.schedules` is a little confusing

firejake308 commented on autoscheduler/frontend/src/redux/reducers/schedules.ts:
```diff
@@ -22,18 +24,77 @@ export interface ReplaceSchedulesAction {
+    case UNSAVE_SCHEDULE:
+      return {
+        ...state,
+        savedSchedules: state.savedSchedules.filter((schedule) => (
```
Why can we use the built-in `filter` here if we have to use a custom `contains` function at line 86?

ryan-conn replied:
The custom `contains` function makes it so if 2 schedules are generated with the same classes in them, only one of them will be saved instead of having 2 identical schedules (which should be a pretty rare edge case), whereas elements in `savedSchedules` are references to elements in `allSchedules`, so checking shallow equality will work

ryan-conn commented:
Now up to date with master, I also changed the UI to the new design that Adel made + addressed his review comments

ryan-conn commented on autoscheduler/frontend/src/components/SchedulingPage/SchedulePreview/SchedulePreview.tsx:
```diff
@@ -83,7 +87,7 @@ const SchedulePreview: React.FC = () => {
       }
     >
-      <List className={styles.list}>
+      <List className={styles.list} disablePadding>
```
Not really related to this PR but the padding on the top and bottom of the list bugged me so I changed it, here's before:
![image](https://user-images.githubusercontent.com/28655462/86315919-190cc080-bbf1-11ea-83b0-088ab47b14a6.png)
and after:
![image](https://user-images.githubusercontent.com/28655462/86315935-2164fb80-bbf1-11ea-9d0d-7d1cbee4f4c5.png)
Let me know if you guys don't like this change, but I think it looks nicer

gannonprudhomme commented on autoscheduler/frontend/src/components/SmallFastProgress.tsx:
```diff
@@ -1,8 +1,18 @@
-const SmallFastProgress: React.FC = () => (
-  <CircularProgress size={24} style={{ animationDuration: '550ms' }} disableShrink />
-);
+interface SmallFastProgressProps {
```
Since this is just going to be a POST/PUT request and the user won't be receiving any data from it (and as such nothing in the view will actually change), I don't think this needs a progress indicator at all & I think it's safe to assume that the request we do worked successfully. Worst case scenario we can show a snackbar in case of an error (although I don't think it's necessary)

ryan-conn replied:
I disagree (at least about removing it entirely), since the most useful part of saving schedules is that after refreshing they'll still be there we want to make sure request to save actually goes through. Imagine that you click save on your perfect schedule, see "saved", refresh, and then it disappears when you refresh. Also because the overhead on the save/remove schedule API calls will probably be very small, I think it's acceptable to have the loading indicator

gannonprudhomme commented on autoscheduler/frontend/src/components/SchedulingPage/SchedulePreview/Buttons/SaveScheduleButton.tsx:
```diff
@@ -0,0 +1,70 @@
+
+  return (
+    <Tooltip title={tooltipText} placement="top">
+      <IconButton className={styles.scheduleButton} size="small" onClick={handleClick} data-testid="save-schedule">
```
Ik `small` is the smallest size MUI provides, but I think the lock & delete icons are still a bit too large - I think they should be shorter than or equal to the schedule text itself

ryan-conn replied:
So apparently icons have a weird behavior where the icon size isn't actually changed by putting `size="small"` on the `IconButton` unless you also put `size="inherit"` on the icon itself, which I didn't notice until you pointed this out. Here's what it looks like now:
![image](https://user-images.githubusercontent.com/28655462/87224909-baf97f00-c34e-11ea-95a8-dafe03ebc427.png)
And here's what it looks like with `size="inherit"` on the buttons:
![image](https://user-images.githubusercontent.com/28655462/87224947-de242e80-c34e-11ea-9e68-02605e431145.png)
Let me know which you like better, or any suggestions for other options

ryan-conn commented:
> ...my main thing that I think is weird about this is the deleting of schedules & how we handle the titles of the schedules. If you delete `Schedule 1`, `Schedule 2` becomes `Schedule 1`, which frankly is pretty confusing. Not sure the best way to deal with this as we might have to completely change how we handle the schedules, but I think if possible we should maintain the indices the schedules received whenever they were generated

I think if we do this, we should just give users the option to change the name of their schedules manually (which shouldn't be too much work, we should be able to just add a `name` prop to schedule and change the schedule name with whatever is in that field).

We should also put that functionality into a separate PR if we decide to implement that functionality, imo it would be a pretty useful feature so it's worth implementing.

Thoughts? @gannonprudhomme @firejake308 @ThatJuanGuy

ThatJuanGuy commented:
I think changing the names of schedules would be a pretty useful feature to have as well. Regarding the indices, I agree with not renaming the remaining schedules when deleting one. It would be pretty confusing, especially if multiple schedules are already saved and they all get renamed.

ryan-conn commented:
Merged in changes from master since we're going to squash and merge PRs now anyways, and addressed feedback so this should be ready for another round of review.

ryan-conn commented:
I removed the loading indicator like we discussed in the previous meeting and made this branch up-to-date with master
